### DISCUSSION TITLE
New rule or `no-floating-promises` upgrade to prevent parallel pending promises before await

### DISCUSSION THREAD

**Discussion by ackvf** [typescript-eslint/typescript-eslint#8097] [RFCs]
### RFC

Given a new rule or [`no-floating-promises`](https://typescript-eslint.io/rules/no-floating-promises/) rule, when an `await` is encountered AND there are more than one pending promises, the rule will suggest to use `await Promise.allSettled` or to append `.catch()` to prevent unexpected behavior that may result in uncaught errors.

### Additional Info

_(Migrated from https://github.com/eslint/eslint/discussions/17824#discussioncomment-7833193)_

In a situation like this, when a promise is awaited (and later fails), while another has already failed, it results in an uncaught error.
In the example below, the failed promise is at the line `const r2 = await p2`, when it throws, the following code never runs and therefore the remaining promises are **never awaited**.

More explanation https://stackoverflow.com/a/69644993/985454

```js
try {
  const p1 = Promise.resolve('p1:success')
  const p2 = Promise.reject('p2:failure')
  const p3 = new Promise((resolve, _) => {
    setTimeout(() => {
      console.log('p3 finished')
      resolve("p3:success")
    }, 3000)
  })
  const p4 = new Promise((_, reject) => {
    setTimeout(() => {
      console.log('p4 finished')
      reject("p4:failure")
    }, 1000)
  })
  

  const r1 = await p1
  const r2 = await p2 // ❌following code never runs!
  const r3 = await p3
  const r4 = await p4
  console.log('await:', [r1, r2, r3, r4])


  // this would also throw ❌ - ✅ caught by no-floating-promises rule
  // Promise.all([p1, p2, p3, p4]).then((values) => {
  //   console.log('all:', values)
  // })


  // this would not throw ✅
  // await Promise.all([p1, p2, p3, p4]).then((values) => {
  //   console.log('all:', values)
  // })


} catch (e) {
  console.log('catch:', e)
}
```

Strangely enough, the `no-floating-promises` rule does not catch these cases
```
const p2 = Promise.reject('p2:failure') // ❌ no error given

const p4 = new Promise((_, reject) => { // ❌ no error given
  setTimeout(() => {
    console.log('p4 finished')
    reject("p4:failure")
  }, 1000)
})

Promise.all([p1, p2, p3, p4]).then((values) => { // ✅ error: Promises must be awaited, end with a call to .catch, ...
  console.log('all:', values)
})
```

### Before you submit your RFC, please confirm the following. If any of these required steps are not taken, we may not be able to review your RFC. Help us to help you!

- [X] I have [read the discussions guidelines](https://typescript-eslint.io/contributing/discussions) and per those guidelines, this fits a discussion category, _not_ a help request or standard issue.
- [X] I have [searched for related discussions](https://github.com/typescript-eslint/typescript-eslint/discussions) and [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

**Comment by bradzacher** [comment]
As per our [discussions contributing guide](https://typescript-eslint.io/contributing/discussions/) (linked in the discussion template) we don't use discussions for rule proposals.

Please file this as an issue.
